modules/abstract-cosmos/src/cosmosCoin.ts

```diff
@@ -127,6 +128,101 @@ export class CosmosCoin extends BaseCoin {
     throw new Error('Method not implemented.');
   }
 
+  private async getMpcV2RecoveryKeyShares(
+    userPublicOrPrivateKeyShare: string,
+    backupPrivateOrPublicKeyShare: string,
+    walletPassphrase?: string
+  ) {
+    const userCompressedPrv = Buffer.from(
+      this.bitgo.decrypt({
+        input: userPublicOrPrivateKeyShare,
+        password: walletPassphrase,
+      }),
+      'base64'
+    );
+    const bakcupCompressedPrv = Buffer.from(
+      this.bitgo.decrypt({
+        input: backupPrivateOrPublicKeyShare,
+        password: walletPassphrase,
+      }),
+      'base64'
+    );
+
+    const userPrvJSON: DklsTypes.ReducedKeyShare = DklsTypes.getDecodedReducedKeyShare(userCompressedPrv);
+    const backupPrvJSON: DklsTypes.ReducedKeyShare = DklsTypes.getDecodedReducedKeyShare(bakcupCompressedPrv);
+    const userKeyRetrofit: DklsTypes.RetrofitData = {
+      xShare: {
+        x: Buffer.from(userPrvJSON.prv).toString('hex'),
+        y: Buffer.from(userPrvJSON.pub).toString('hex'),
+        chaincode: Buffer.from(userPrvJSON.rootChainCode).toString('hex'),
+      },
+      xiList: userPrvJSON.xList.slice(0, 2),
+    };
+    const backupKeyRetrofit: DklsTypes.RetrofitData = {
+      xShare: {
+        x: Buffer.from(backupPrvJSON.prv).toString('hex'),
+        y: Buffer.from(backupPrvJSON.pub).toString('hex'),
+        chaincode: Buffer.from(backupPrvJSON.rootChainCode).toString('hex'),
+      },
+      xiList: backupPrvJSON.xList.slice(0, 2),
+    };
+    const [user, backup] = await DklsUtils.generate2of2KeyShares(userKeyRetrofit, backupKeyRetrofit);
+    const userKeyShare = user.getKeyShare();
+    const backupKeyShare = backup.getKeyShare();
+    const commonKeyChain = DklsTypes.getCommonKeychain(userKeyShare);
+    const MPC = new Ecdsa();
+    const derivedCommonKeyChain = MPC.deriveUnhardened(commonKeyChain, 'm');
+    const backupKeyPair = new KeyPairLib({ pub: derivedCommonKeyChain.slice(0, 66) });
+    const baseAddress = backupKeyPair.getAddress();
+    return { userKeyShare, backupKeyShare, commonKeyChain, baseAddress };
+  }
+
+  private static async signRecoveryMpcV2(
```

Comment:
I think we can move this too to DklsUtils. We need to change the function signature, instead of `tx: CosmosTransaction` take messageHash as input